core/client/src/main/java/alluxio/client/block/RetryHandlingBlockWorkerClient.java

```diff
@@ -112,7 +126,12 @@ public void releaseClient(BlockWorkerClientService.Client client) {
   @Override
   public void close() {
     if (mHeartbeat != null) {
-      mHeartbeat.cancel(true);
+      HEARTBEAT_CANCEL_POOL.submit(new Runnable() {
+        @Override
+        public void run() {
+          mHeartbeat.cancel(true);
+        }
+      });
```

Comment:
It is in separate threadpool to avoid deadlock. If all the threads for heatbeat is consumed by heatbeats, no one can be cancelled if we try to run cancel in that threadpool